### ISSUE TITLE
Scripts share global namespace

### ISSUE THREAD

**Issue by maggu2810** [eclipse/smarthome#3008]
Do scripts (of different rules) share their global namespace by intention?

I created two rules.
Every rule contains an action that executes a JavaScript script.

This is the script of the first rule's action:

```JavaScript
try {
  myvariable;
  print("Variable is declared");
  if (myvariable !== undefined) {
    if (myvariable !== null) {
      print("The variable's value is: " + myvariable)
    } else {
      print("The variable's value is null");
    }
  } else {
    print("The variable's value is undefined");
  }
}
catch(e) {
  if (e.name == "ReferenceError") {
    print("Variable is not declared");
  }
}
```

If the script (the first one) is executed it looks like:

```text
16:00:13.047 INFO  s.event.RuleStatusInfoEvent[:43] - rule_2 updated: RUNNING
Variable is not declared
16:00:13.051 DEBUG o.e.s.a.c.internal.RuleEngine[:883] - The rule 'rule_2' is executed.
16:00:13.053 INFO  s.event.RuleStatusInfoEvent[:43] - rule_2 updated: IDLE
```

This is the script of the second rule's action:

```JavaScript
myvariable = 2;
```

If the script (the second one) is executed it looks like:

```text
16:01:19.608 INFO  s.event.RuleStatusInfoEvent[:43] - rule_3 updated: RUNNING
16:01:19.639 DEBUG o.e.s.a.c.internal.RuleEngine[:883] - The rule 'rule_3' is executed.
16:01:19.641 INFO  s.event.RuleStatusInfoEvent[:43] - rule_3 updated: IDLE
```

If the first script is executed again, it looks like:

```text
Variable is declared
16:02:01.663 INFO  s.event.RuleStatusInfoEvent[:43] - rule_2 updated: RUNNING
The variable's value is: 2
16:02:01.664 DEBUG o.e.s.a.c.internal.RuleEngine[:883] - The rule 'rule_2' is executed.
16:02:01.665 INFO  s.event.RuleStatusInfoEvent[:43] - rule_2 updated: IDLE
```

I would expect that every script uses its own global namespace or at least scripts of different rules.
Has this been designed this way by intention?

IMHO this way there is a magic pipe how scripts can share data and the script logic could depend on execution order of other ones etc.

**Comment by kaikreuzer** [comment]
> Has this been designed this way by intention?

No, I don't think so. If I am not mistaken, this is also fixed by https://github.com/eclipse/smarthome/pull/1783 - maybe @smerschjohann can comment as well?

**Comment by smerschjohann** [comment]
Yes, this should be fixed right now and also in the future.

Currently, I create a new ScriptEngine instance each time I'm executing the Handler. That would mean that no state can be preserved at all. I will change that to each time I'm creating a handler instance, this will allow the script to keep state in the context of the rule.

**Comment by maggu2810** [comment]
Fixed by: #1783
